### ISSUE TITLE
py_compat test is "too slow"

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#381]
Originally reported of Google Code with ID 325.

```
More than one minute, 29 cpu seconds, and 99% cpu... can we fix this?

$ time htest/test -t OpCode
OpCodes:
  serialization: [OK, passed 500 tests]
  AllDefined: [OK]
  py_compat: [OK]

         Properties  Test Cases  Total      
 Passed  1           2           3          
 Failed  0           0           0          
 Total   1           2           3          

real	1m13.492s
user	0m29.690s
sys	0m2.604s


```
Originally added on 2012-11-30 17:51:03 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
There is something wrong there. The test "ate" 30 seconds of CPU time, but it took in total more than a minute? On my machine, it's real 30.738s, user 29.265s. Is something wrong with starting Python, or was your machine actually very busy?

Also, we have other tests which take a long time. The only problem that I see is that the test doesn't show output on the screen while doing this, so it looks like "it's stuck" :)

I'll take a look to see if we generate bad data, but given that we generate, serialise, pass to python, deserialise, validate, serialise, pass back to Haskell 38,500 opcodes (500 * count(opcodes)), it will be slow.

We can also reduce the number of opcodes (e.g. 100 per type), which would make the test faster.
```
Originally added on 2012-12-01 13:44:43 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I double checked and I believe the problem is a memory one: that command seems to be using up to 2GB resident memory towards the end, which on the system where I run it are not available, and thus cause the system to swap and the command to be that slow.

Perhaps reducing the number of opcodes run to 100 can help, or perhaps we can find some other way to aggressively free memory?


Thanks,

Guido
```
Originally added on 2012-12-01 15:16:39 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
(Running the command with -s 100 didn't produce any improvement, memory-wise)

Guido
```
Originally added on 2012-12-01 15:19:54 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
> Perhaps reducing the number of opcodes run to 100 can help, or perhaps we can find some other way to aggressively free memory?

Ah, I see. Yes, that is a problem indeed.

Even before your report, I was thinking that "create N opcodes, pass them in bulk to Python, get them back and check" is not nice, and that we should have a way to stream opcodes back and forth (e.g. in groups of 1000) or so. But that would require us to use a luxi-like protocol over stdin/stdout in Python and Haskell, so it would need some work…

I'll take a look, most probably this doesn't need that much memory.

> Running the command with -s 100 didn't produce any improvement, memory-wise

Yes, because - if you look at the code of the test - the '500' is hardcoded and doesn't come from the -s option for this test.

Thanks!
```
Originally added on 2012-12-02 12:52:29 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
Just for the record, this is what I get when running just that test:

$ ./htest/test -t OpCodes/py_compat +RTS -s
OpCodes:
  py_compat: [OK]

         Test Cases  Total      
 Passed  1           1          
 Failed  0           0          
 Total   1           1          
  32,990,909,040 bytes allocated in the heap
   4,668,968,424 bytes copied during GC
     809,045,280 bytes maximum residency (15 sample(s))
       8,208,192 bytes maximum slop
            2008 MB total memory in use (0 MB lost due to fragmentation)

                                    Tot time (elapsed)  Avg pause  Max pause
  Gen  0     64007 colls,     0 par    3.93s    3.94s     0.0001s    0.0015s
  Gen  1        15 colls,     0 par    3.77s    3.83s     0.2552s    1.3856s

  INIT    time    0.00s  (  0.04s elapsed)
  MUT     time   17.80s  ( 22.54s elapsed)
  GC      time    7.70s  (  7.76s elapsed)
  EXIT    time    0.07s  (  0.11s elapsed)
  Total   time   25.60s  ( 30.49s elapsed)

  %GC     time      30.1%  (25.5% elapsed)

  Alloc rate    1,853,560,984 bytes per MUT second

  Productivity  69.9% of total user, 58.7% of total elapsed

Not only it uses lots of memory, but also trashes the GC somewhat (30% of time spent in the GC).
```
Originally added on 2012-12-02 12:57:43 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
From some quick time profiling, we have:

COST CENTRE                 MODULE                 %time %alloc  ticks     bytes

arbitrary                   Test.Ganeti.TestCommon  39.0   31.9   1581 1172646208
genName                     Test.Ganeti.TestCommon  27.7   34.8   1124 1276237896

and from space profiling, it seems that we have two phases of:

- first, generation of the opcodes, which eats some memory
- then, after sending them to python, reading them, which causes (incomplete deserialisation? lazyness?) even more memory usage

The bad news is that looking at the biography profiling, none of the memory is wasted (as in, generated but not ever looked at). So this looks like the actual memory consumption of the damn String type (all the memory is used, when looking at the closure/type descriptions, in lists, which sounds like Strings indeed).

For fun, I've attached the various memory profiling graphs - you can see the two phases there easily.

Did I mention I can't wait to move away from String+json to ByteString/Text+aeson? Waiting for wheezy to be released, and then we can start looking at the migration…

```
Originally added on 2012-12-02 14:17:27 +0000 UTC.

Attachments: 
[memory.pdf](https://storage.googleapis.com/google-code-attachments/ganeti/issue-325/comment-6/memory.pdf)

**Comment by Ganeti-Issues-Migrator** [comment]
```
One more note: changing the maximum length of auto-generated names from 64 to 8 (times 4, both; this is about each individual component in an FQDN) reduces the memory usage by around 3 times. So yes, this is definitely related to Strings being used for the various elements in the opcodes (hostnames, etc.).

The other part relatd to strings is the fact that the encoded (sent to python) and decoded (received from python) structure is again strings, whereas JSON encoded is actually bytestrings, so we lose memory there as well.

In summary, for now we can do:

- reduce the number of opcodes from 500 * count(opcodes) to 100 * count()
- reduce the hostname component size from 64 to 16

This makes the runtime and memory consumption reduce by 10 times, so about 3-4 seconds of CPU usage and ~200 MB of memory. How does this sound?
```
Originally added on 2012-12-02 17:04:01 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Thanks, this would be a lot better!

Guido
```
Originally added on 2012-12-02 18:27:16 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
OK, patch submitted.
```
Originally added on 2012-12-03 14:11:22 +0000 UTC.

Changed State: Released
